addons/binding/org.openhab.binding.smappee/src/main/java/org/openhab/binding/smappee/internal/SmappeeService.java

```diff
@@ -0,0 +1,473 @@
+import java.util.concurrent.TimeoutException;
+
+import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.jetty.client.api.ContentResponse;
+import org.eclipse.jetty.client.api.Request;
+import org.eclipse.jetty.client.util.FormContentProvider;
+import org.eclipse.jetty.client.util.StringContentProvider;
+import org.eclipse.jetty.http.HttpMethod;
+import org.eclipse.jetty.http.HttpStatus;
+import org.eclipse.jetty.util.Fields;
+import org.eclipse.jetty.util.ssl.SslContextFactory;
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonSyntaxException;
+
+/**
+ * The {@link SamppeeService} handles the connection to the Smappee API
+ *
+ * @author Niko Tanghe - Initial contribution
+ */
+public class SmappeeService {
+
+    private final Logger logger = LoggerFactory.getLogger(SmappeeService.class);
+
+    ScheduledFuture<?> scheduledJob;
+
+    private int retry;
+
+    public SmappeeConfigurationParameters config;
+
+    private String serviceLocationId;
+
+    private boolean initialized;
+
+    private final Gson gson;
+
+    private String accessToken;
+    private DateTime accessTokenValidity;
+    private String refreshToken;
+
+    private static HttpClient httpClient = new HttpClient(new SslContextFactory());
+
+    public SmappeeService(SmappeeConfigurationParameters config) {
+        this.config = config;
+
+        this.retry = 1;
+
+        this.gson = new Gson();
+
+        this.initialized = false;
+
+        if (!httpClient.isStarted()) {
+            try {
+                httpClient.setFollowRedirects(false);
+                httpClient.start();
+            } catch (Exception e) {
+                logger.warn("Cannot start HttpClient!", e);
+            }
+        }
+
+    }
+
+    public void startAutomaticRefresh(ScheduledExecutorService scheduledExecutorService,
+            final ReadingsUpdate readingsUpdate) {
+        scheduledJob = scheduledExecutorService.scheduleWithFixedDelay(() -> {
+            SmappeeDeviceReading readings = getDeviceReadings();
+
+            // sometimes smappee returns reading without aggregated consumptions
+            if (readings != null && readings.consumptions.length > 0) {
+                readingsUpdate.newState(readings);
+            }
+        }, 0, config.poll_time, TimeUnit.MINUTES);
+    }
+
+    public void stopAutomaticRefresh() {
+        if (scheduledJob != null) {
+            scheduledJob.cancel(true);
+        }
+    }
+
+    /**
+     * Gets the smappee energy consumption readings.
+     *
+     * @return the device readings
+     */
+    public SmappeeDeviceReading getDeviceReadings() {
+        int currentTry = 0;
+        do {
+            try {
+                DateTime nowUtc = DateTime.now(DateTimeZone.UTC).minusMinutes(1);
+                DateTime nowUtcMinus20Min = DateTime.now(DateTimeZone.UTC).minusMinutes(20);
```

Comment:
Isn't "the last 19 Minutes" a bit unintuitive? Wouldn't it make more sense to have something widely known as a default (e.g. 24h) and let the user configure this?

Reply:
@sjka What do you mean?

Reply:
What are those 19 minutes about? Sounds arbitrary to me without having any further background here...